nifi-nar-bundles/nifi-druid-bundle/nifi-druid-processors/src/main/java/org/apache/nifi/processors/PutDruid.java

```diff
@@ -0,0 +1,206 @@
+         //Convert each array element from JSON to HashMap and send to Druid
+         for(String message: messageArray){
+         	try {
+         		contentMap = new ObjectMapper().readValue(message, HashMap.class);
+         		//contentMap = new ObjectMapper().readValue(message, HashMap.class);
+         	} catch (JsonParseException e) {
+         		getLogger().error("Error parsing incoming message array in the flowfile body");
+         	} catch (JsonMappingException e) {
+         		getLogger().error("Error parsing incoming message array in the flowfile body");
+         	} catch (IOException e) {
+         		getLogger().error("Error parsing incoming message array in the flowfile body");
+         	}
+
+         	getLogger().debug("Tranquilizer Status: " + tranquilizer.status().toString());
+         	messageStatus.put(flowFile, "pending");
+         	//Send data element to Druid, Asynch
+         	Future<BoxedUnit> future = tranquilizer.send(contentMap);
+         	getLogger().debug(" Sent Payload to Druid: " + contentMap);
+         
+         	//Wait for Druid to call back with status 
+         	future.addEventListener(new FutureEventListener<Object>() {
+         		@Override
+         		public void onFailure(Throwable cause) {
+         			if (cause instanceof MessageDroppedException) {
+         				//This happens when event timestamp targets a Druid Indexing task that has closed (Late Arriving Data) 
+         				getLogger().error(" FlowFile Dropped due to MessageDroppedException: " + cause.getMessage() + " : " + cause);
+         				cause.getStackTrace();
+         				getLogger().error(" Transfering FlowFile to DROPPED relationship");
+         				session.transfer(flowFile, REL_DROPPED);
+         			} else {
+         				getLogger().error(" FlowFile Processing Failed due to: " + cause.getMessage() + " : " + cause);
+         				cause.printStackTrace();
+         				getLogger().error(" Transfering FlowFile to FAIL relationship");
+         				session.transfer(flowFile, REL_FAIL);
+         			}
+         		}
+
+         		@Override
+         		public void onSuccess(Object value) {
+         			getLogger().debug(" FlowFile Processing Success : "+ value.toString());
+         			session.transfer(flowFile, REL_SUCCESS);
+         		}
+         	});
+
+         	try {
+         		//Wait for result from Druid 
+         		//This method will be asynch since this is a SessionFactoryProcessor and OnTrigger will create a new Thread
+         		Await.result(future);
+         	} catch (Exception e) {
+         		e.printStackTrace();
+         	}
+         }	
+         //session.transfer(flowFile, REL_SUCCESS);
+         session.commit();
+    }
+    
+    public void onTrigger(ProcessContext context, ProcessSessionFactory factory) throws ProcessException {
+    	final ProcessSession session = factory.createSession();
+    	//Create new Thread to ensure that waiting for callback does not reduce throughput
+    	new Thread(() -> {
```

Comment:
It does not appear that spawning a new thread is necessary and it is unclear that is desirable.  If more throughput is needed we would add threads to the processor or use the API of druid differently etc..  But we don't want to spawn a thread per onTrigger/session

Reply:
The problem is we need to wait for the Future to come back so that we can send the flow file to the correct relationship (SUCCESS, DROPPED, FAIL). This means that we have to Await.result(future);. If we do that but don't make each onTrigger Asynch, the throughput comes to a crawl. If we don't block while waiting for Future call back, we have to assume SUCCESS and thus can't handle late arriving data. Won't the thread close when commit is called? Combined with the fact that this is an AbstractSessionFactoryProcessor, should act as a kind of self managing thread pool. If not, what is the best way to make onTrigger Asynch without spinning threads? Is there some annotation magic I should use.